### ISSUE TITLE
Cookies headers is present but Cookies are not stored in browser

### ISSUE THREAD

**Issue by Tarasovych** [axios/axios#1553]
I'm making cross-domain request between two subdomains: `sub1.local` and `sub2.local`. Both are on local, as you see.

So I send `POST` from `sub1.local` to `sub2.local`.
I recevie Cookie header: `Set-Cookie: name=value; expires=Sun, 20-May-2018 21:38:08 GMT; Max-Age=3600; path=/; domain=.local`.
But there is no Cookie in browser storage.

All response headers:
```
HTTP/1.1 200 OK
Date: Sun, 20 May 2018 20:43:05 GMT
Server: Apache
Set-Cookie: name=value; expires=Sun, 20-May-2018 21:43:05 GMT; Max-Age=3600; path=/; domain=.local
Cache-Control: no-cache, private
Access-Control-Allow-Origin: http://sub1.local:8080
Vary: Origin
X-RateLimit-Limit: 60
X-RateLimit-Remaining: 59
Content-Length: 2
Keep-Alive: timeout=10, max=100
Connection: Keep-Alive
Content-Type: text/html; charset=UTF-8
```

What can be wrong?

I've tried add `withCredentials: true` after reading [this question](https://stackoverflow.com/questions/22432616/why-is-the-browser-not-setting-cookies-after-an-ajax-request-returns) - didn't helped.

**Comment by jeromemalot** [comment]
Hello @Tarasovych ,
I think you need to indicate whether or not the response to the request can be exposed to the page.
Could you try to add the following middleware on your Express backend API (sub2.local).

```
app.use(function(req, res, next) {
  res.header('Access-Control-Allow-Credentials', true);
  res.header('Access-Control-Allow-Origin', req.headers.origin);
  res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE,UPDATE,OPTIONS');
  res.header('Access-Control-Allow-Headers', 'X-Requested-With, X-HTTP-Method-Override, Content-Type, Accept');
  next();
});
```
Once in production, don't forget to modify req.headers.origin to the exact website you wish to allow to connect.

**Comment by Tarasovych** [comment]
@jeromemalot Hello
I'm using Laravel as backend API.
I'm also using [laravel-cors](https://github.com/barryvdh/laravel-cors) for handling cross-domain requests. At the moment all methods are allowed, credentials are `true`.

If backend is able to send Cookie but Vue app doesn't accept them, there might be a problem not with Vue but with backend?

**Comment by jeromemalot** [comment]
_I'm also using laravel-cors for handling cross-domain requests. At the moment all methods are allowed, credentials are true._
=>Did you also allowed origins? (and headers to anticipate issues)
`'allowedOrigins' => ['*'],
'allowedHeaders' => ['Content-Type', 'X-Requested-With']
`
_If backend is able to send Cookie but Vue app doesn't accept them, there might be a problem not with Vue but with backend?_
=>Agree hence my answer, your backend needs to indicate whether or not the response is exposed. Technically the blocker is your browser (if you test it on Postman, I think it should works... good test to complete btw). That's why your backend needs to play with your browser's rules.

**Comment by Tarasovych** [comment]
@jeromemalot yes, I've allowed origins and all headers are allowed too. My current cors config:
```
    'supportsCredentials' => true,
    'allowedOrigins' => ['*'],
    'allowedOriginsPatterns' => [],
    'allowedHeaders' => ['*'],
    'allowedMethods' => ['*'],
    'exposedHeaders' => [],
    'maxAge' => 0,
```
I've tested in Postamn analog (Insomnia) - everything was good...

**Comment by jeromemalot** [comment]
... weird. If Insomnia works well (I mean you receive cookies) and you can't see your cookies on your browser, my 2cents is that we should looking for a backend fix. I'm not familiar with Laravel but I would try doing some fine tuning with the cors config. Sorry can't help further :(

**Comment by Tarasovych** [comment]
I've found an answer on stack about the problem that Chrome won't set Cookie if there is port in domain url. Can I remove port from domain url? I'm using vue-cli.

**Comment by Mont4** [comment]
@Tarasovych did you find a solution for it?

**Comment by Tarasovych** [comment]
@Mont4 can't answer unequivocally. Sometimes I have just restarted server and it worked...

And also [your request might have `withCredentials = true` flag](https://developer.mozilla.org/uk/docs/Web/HTTP/CORS#Requests_with_credentials). Can be various implementations of this, depending on what tool do you use to make AJAX requests.

**Comment by Tarasovych** [comment]
Note that `withCredentials` flag might be not in the `data` parameter:
```js
  axios.post('domain.com', {
    //data parameter
    withCredentials: true, //wrong
    name: 'name',
    password: 'password',
  })
```
but
```js
  axios.post('domain.com', {
    name: 'name',
    password: 'password'
  }, {
    //AxiosRequestConfig parameter
    withCredentials: true //correct
  })
```

**Comment by sajadreshi** [comment]
I faced the same issue in chrome. My url is localhost:3000. Looks like chrome doesn't like when port is there in the url. However, I tested with Forfox and it worked fine.

**Comment by Skinner927** [comment]
For any poor SOB who lands on this page as I did, `127.0.0.1` != `localhost`. If you are serving your site from `localhost`(port whatever) and making CORS requests to `127.0.01` (port whatever) cookies will not stick! Ensure to use either only `localhost` or only `127.0.0.1`, but do not mix and match.

**Comment by chaims** [comment]
same problem

**Comment by bos-hieu** [comment]
how to fix this issue?

**Comment by jplps** [comment]
> Note that `withCredentials` flag might be not in the `data` parameter:
> 
> ```js
>   axios.post('domain.com', {
>     //data parameter
>     withCredentials: true, //wrong
>     name: 'name',
>     password: 'password',
>   })
> ```
> 
> but
> 
> ```js
>   axios.post('domain.com', {
>     name: 'name',
>     password: 'password'
>   }, {
>     //AxiosRequestConfig parameter
>     withCredentials: true //correct
>   })
> ```

MAGIC! Thanks a LOT little crustacean!

**Comment by MOKuper** [comment]
Changing the SPA origin domain fixed it for me.

I was making a SPA request to a Laravel backend, and the origin of the request (SPA) was 127.0.0.1. I found that adding a vue.config.js file in the root of the project (level of package.json) with the snipped below, I was able to align the domains. (SESSION_DOMAIN=.127.0.0.1 in the Laravel env file didn't work for me).
```
module.exports = {
  devServer: {
    host: 'projectname.local'
  }
}
```

my source; https://stevencotterill.com/snippets/vue-cli-using-a-custom-domain
